In [1]:
import pandas as pd
from google.colab import files
import io

In [2]:
import numpy as np
import matplotlib.pyplot as plt

In [3]:
uploaded = files.upload()
df = pd.read_csv(io.BytesIO(uploaded["data-breast-cancer.csv"]))

Saving data-breast-cancer.csv to data-breast-cancer.csv


In [4]:
df.head()

Unnamed: 0 diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0           0         M        17.99         10.38          122.80     1001.0   
1           1         M        20.57         17.77          132.90     1326.0   
2           2         M        19.69         21.25          130.00     1203.0   
3           3         M        11.42         20.38           77.58      386.1   
4           4         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  fractal_dimension_mean  
0         0.2419                 0.07871  
1         0.1812                 0.05667  
2         0.2069                 0.05999  
3         0.2597                 0.09744  
4         0.1809                 0.05883

# EXPLORING DATA

In [5]:
df.info()
df.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Unnamed: 0              569 non-null    int64  
 1   diagnosis               569 non-null    object 
 2   radius_mean             569 non-null    float64
 3   texture_mean            569 non-null    float64
 4   perimeter_mean          569 non-null    float64
 5   area_mean               569 non-null    float64
 6   smoothness_mean         569 non-null    float64
 7   compactness_mean        569 non-null    float64
 8   concavity_mean          569 non-null    float64
 9   concave points_mean     569 non-null    float64
 10  symmetry_mean           569 non-null    float64
 11  fractal_dimension_mean  569 non-null    float64
dtypes: float64(10), int64(1), object(1)
memory usage: 53.5+ KB


(569, 12)

In [6]:
# Drop duplicate samples
df = df.drop_duplicates(ignore_index=True)
df.shape

(569, 12)

In [7]:
#Remove outliers of features
q = df['radius_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['radius_mean'] < q]


q = df['texture_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['texture_mean'] < q]


q = df['perimeter_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['perimeter_mean'] < q]

q = df['area_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['area_mean'] < q]


q = df['smoothness_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['smoothness_mean'] < q]

q = df['compactness_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['compactness_mean'] < q]


q = df['concavity_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['concavity_mean'] < q]

q = df['concave points_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['concave points_mean'] < q]


q = df['symmetry_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['symmetry_mean'] < q]


q = df['fractal_dimension_mean'].quantile(0.98)  # Select q range as 98%
df_clean = df[df['fractal_dimension_mean'] < q]


In [8]:
df_clean.shape

(557, 12)

In [58]:
from sklearn.preprocessing import LabelEncoder
label = LabelEncoder()
label.fit(df_clean['diagnosis'])
df_clean['diagnosis'] = label.transform(df_clean['diagnosis'])
df_clean

Unnamed: 0  diagnosis  radius_mean  texture_mean  perimeter_mean  \
0             0          1        17.99         10.38          122.80   
1             1          1        20.57         17.77          132.90   
2             2          1        19.69         21.25          130.00   
4             4          1        20.29         14.34          135.10   
5             5          1        12.45         15.70           82.57   
..          ...        ...          ...           ...             ...   
564         564          1        21.56         22.39          142.00   
565         565          1        20.13         28.25          131.20   
566         566          1        16.60         28.08          108.30   
567         567          1        20.60         29.33          140.10   
568         568          0         7.76         24.54           47.92   

     area_mean  smoothness_mean  compactness_mean  concavity_mean  \
0       1001.0          0.11840           0.27760         0.30010   
1       1326.0          0.08474           0.07864         0.08690   
2       1203.0          0.10960           0.15990         0.19740   
4       1297.0          0.10030           0.13280         0.19800   
5        477.1          0.12780           0.17000         0.15780   
..         ...              ...               ...             ...   
564     1479.0          0.11100           0.11590         0.24390   
565     1261.0          0.09780           0.10340         0.14400   
566      858.1          0.08455           0.10230         0.09251   
567     1265.0          0.11780           0.27700         0.35140   
568      181.0          0.05263           0.04362         0.00000   

     concave points_mean  symmetry_mean  fractal_dimension_mean  
0                0.14710         0.2419                 0.07871  
1                0.07017         0.1812                 0.05667  
2                0.12790         0.2069                 0.05999  
4                0.10430         0.1809                 0.05883  
5                0.08089         0.2087                 0.07613  
..                   ...            ...                     ...  
564              0.13890         0.1726                 0.05623  
565              0.09791         0.1752                 0.05533  
566              0.05302         0.1590                 0.05648  
567              0.15200         0.2397                 0.07016  
568              0.00000         0.1587                 0.05884  

[557 rows x 12 columns]

# SVM

In [59]:
# Separate data features by removing the data label.
X = df_clean.drop(columns=["diagnosis"], axis=1)

# Assign data label to variable y
y = df_clean.diagnosis

# Split train/test with a random state
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=10, train_size=0.8)

In [52]:
# Show some training samples
X_train.head()

Unnamed: 0  radius_mean  texture_mean  perimeter_mean  area_mean  \
401         401        11.93         10.91           76.14      442.7   
342         342        11.06         14.96           71.49      373.9   
179         179        12.81         13.06           81.29      508.8   
400         400        17.91         21.02          124.40      994.0   
92           92        13.27         14.76           84.74      551.7   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
401          0.08872           0.05242        0.026060              0.01796   
342          0.10330           0.09097        0.053970              0.03341   
179          0.08739           0.03774        0.009193              0.01330   
400          0.12300           0.25760        0.318900              0.11980   
92           0.07355           0.05055        0.032610              0.02648   

     symmetry_mean  fractal_dimension_mean  
401         0.1601                 0.05541  
342         0.1776                 0.06907  
179         0.1466                 0.06133  
400         0.2113                 0.07115  
92          0.1386                 0.05318

In [60]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 557 entries, 0 to 568
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Unnamed: 0              557 non-null    int64  
 1   diagnosis               557 non-null    int64  
 2   radius_mean             557 non-null    float64
 3   texture_mean            557 non-null    float64
 4   perimeter_mean          557 non-null    float64
 5   area_mean               557 non-null    float64
 6   smoothness_mean         557 non-null    float64
 7   compactness_mean        557 non-null    float64
 8   concavity_mean          557 non-null    float64
 9   concave points_mean     557 non-null    float64
 10  symmetry_mean           557 non-null    float64
 11  fractal_dimension_mean  557 non-null    float64
dtypes: float64(10), int64(2)
memory usage: 56.6 KB


In [61]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_normalized_train = scaler.fit_transform(X_train)     # Fit and transform thr training data
X_normalized_test = scaler.transform(X_test)

In [62]:
from sklearn.svm import SVC
model = SVC()
model.fit(X_normalized_train, y_train)

SVC()

In [63]:
# Show evaluation metrics on the test set
from sklearn.metrics import classification_report
print(classification_report(y_test, model.predict(X_normalized_test)))

              precision    recall  f1-score   support

           0       0.96      0.96      0.96        73
           1       0.92      0.92      0.92        39

    accuracy                           0.95       112
   macro avg       0.94      0.94      0.94       112
weighted avg       0.95      0.95      0.95       112



Choose best set of parameters

In [64]:
from sklearn.model_selection import GridSearchCV
param_grid = {"C": [0.01, 0.1, 1, 10, 100, 1000],
              "gamma": ["scale", 0.001, 0.005, 0.1]}
gridsearch = GridSearchCV(SVC(), param_grid, cv=10, scoring="f1", verbose=1) 

In [65]:
gridsearch.fit(X_normalized_train, y_train)

Fitting 10 folds for each of 24 candidates, totalling 240 fits


GridSearchCV(cv=10, estimator=SVC(),
             param_grid={'C': [0.01, 0.1, 1, 10, 100, 1000],
                         'gamma': ['scale', 0.001, 0.005, 0.1]},
             scoring='f1', verbose=1)

In [66]:
gridsearch.best_params_

{'C': 10, 'gamma': 'scale'}

In [67]:
# Re-run SVM with the best set of hyper-parameters.
model = SVC(C=gridsearch.best_params_['C'], gamma=gridsearch.best_params_['gamma'])
model.fit(X_normalized_train, y_train)

SVC(C=10)

In [68]:
# Show evaluation metrics on the test set
print(classification_report(y_test, model.predict(X_normalized_test)))

              precision    recall  f1-score   support

           0       0.96      0.95      0.95        73
           1       0.90      0.92      0.91        39

    accuracy                           0.94       112
   macro avg       0.93      0.93      0.93       112
weighted avg       0.94      0.94      0.94       112



# LOGISTIC REGRESSION

In [69]:
y = df_clean['diagnosis']                                        # Initialize label y

# Generate feature X from data frame
X = df_clean.drop('diagnosis', axis=1, inplace=False).values 

In [70]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)         # Train-test split with fixed random state


In [71]:
from sklearn.preprocessing import StandardScaler     # Import function to perform Standardization
scaler = StandardScaler()                            # Initialize the StandardScaler
X_train_scaler = scaler.fit_transform(X_train)       # Fit (compute mean and std), then transform training data using these values.
X_test_scaler = scaler.transform(X_test)

In [72]:
from sklearn.linear_model import LogisticRegression
logmodel = LogisticRegression()                      # Initialize Logistic Regression model
logmodel.fit(X_train_scaler, y_train) 

LogisticRegression()

In [73]:
# Calculate test accuracy
logmodel.score(X_test_scaler, y_test)

0.9553571428571429

In [74]:
# Calculate prediction labels
predictions = logmodel.predict(X_test_scaler)                # Predict labels for X_test_scaler
final_prediction = pd.DataFrame({'target':predictions})      # Import labels to a dataframe 
print(final_prediction)

     target
0         0
1         0
2         0
3         1
4         0
..      ...
107       0
108       1
109       0
110       1
111       1

[112 rows x 1 columns]


In [75]:
from sklearn.metrics import precision_score, recall_score, f1_score, confusion_matrix, classification_report

print("Precision: ", precision_score(logmodel.predict(X_test_scaler), y_test))
print("Recall: ", recall_score(logmodel.predict(X_test_scaler), y_test))
print("F1: ", f1_score(logmodel.predict(X_test_scaler), y_test))
print("Confusion matrix: \n", confusion_matrix(logmodel.predict(X_test_scaler), y_test))
print("Classification report: \n", classification_report(logmodel.predict(X_test_scaler), y_test))

Precision:  0.92
Recall:  0.9787234042553191
F1:  0.9484536082474226
Confusion matrix: 
 [[61  4]
 [ 1 46]]
Classification report: 
               precision    recall  f1-score   support

           0       0.98      0.94      0.96        65
           1       0.92      0.98      0.95        47

    accuracy                           0.96       112
   macro avg       0.95      0.96      0.95       112
weighted avg       0.96      0.96      0.96       112



# NAIVE BAYES

In [76]:
from sklearn.naive_bayes import GaussianNB
naive_model = GaussianNB()
naive_model.fit(X_train_scaler, y_train)

GaussianNB()

In [77]:
from sklearn.metrics import precision_score, recall_score, f1_score, classification_report

# Make prediction on the test data
predicted_label = naive_model.predict(X_test_scaler)

# Calculate evaluation metrics by comparing the prediction with the data label y_test
print(precision_score(predicted_label, y_test))
print(recall_score(predicted_label, y_test))
print(f1_score(predicted_label, y_test))
print(classification_report(predicted_label, y_test))

0.9
0.9574468085106383
0.9278350515463918
              precision    recall  f1-score   support

           0       0.97      0.92      0.94        65
           1       0.90      0.96      0.93        47

    accuracy                           0.94       112
   macro avg       0.93      0.94      0.94       112
weighted avg       0.94      0.94      0.94       112



# DECISION TREE

In [79]:
from imblearn.over_sampling import SMOTE     # Load the SMOTE library
smote = SMOTE(random_state=5)                # Initialize SMOTE
X_train_oversampling, y_train_oversampling = smote.fit_resample(X_train, y_train) 

In [78]:
print("Before oversampling: " + str(X_train.shape))

Before oversampling: (445, 11)


In [81]:
X_train_oversampling.shape

(574, 11)

In [82]:
# Import GridSearchCV for finding the best hyper-parameter set.
from sklearn.pipeline import make_pipeline
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor

params = {"criterion": ["gini", "entropy"],             # Criterion to evaluate the purity.
         "max_depth": [3, 5],                           # Maximum depth of the tree
         "min_samples_split": [4, 8]}                   # Stop splitting condition.

grid_search = GridSearchCV(estimator=DecisionTreeClassifier(), param_grid=params, cv=5)

In [83]:
# Run the search on oversampled training data samples.
grid_search.fit(X_train_oversampling, y_train_oversampling)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'], 'max_depth': [3, 5],
                         'min_samples_split': [4, 8]})

In [84]:
# Best set of hyper-parameters found after searching
grid_search.best_params_

{'criterion': 'entropy', 'max_depth': 5, 'min_samples_split': 4}

In [85]:
# Build a decision tree model pipeline from the best set of hyper-parameters found
model_dt = DecisionTreeClassifier(criterion=grid_search.best_params_['criterion'], max_depth=grid_search.best_params_['max_depth'], min_samples_split=grid_search.best_params_['min_samples_split'])

In [86]:
# Train the decision tree model
model_dt.fit(X_train_oversampling, y_train_oversampling)

DecisionTreeClassifier(criterion='entropy', max_depth=5, min_samples_split=4)

In [87]:
# Calculate some evaluation metrics: precision, recall, f1-score. 'Support' is the number of actual occurrences of the class in the dataset.
from sklearn.metrics import classification_report

# Make prediction on the original test set (after training on the over-sampled training set).
pred_y = model_dt.predict(X_test)

print(classification_report(y_test, pred_y))

              precision    recall  f1-score   support

           0       0.92      0.94      0.93        62
           1       0.92      0.90      0.91        50

    accuracy                           0.92       112
   macro avg       0.92      0.92      0.92       112
weighted avg       0.92      0.92      0.92       112



# RANDOM FOREST

In [88]:
# Building a RandomForest is similar to building a decision tree, but there is an additional hyper-parameter called n_estimators which is the number of trees in the forest.
from sklearn.ensemble import RandomForestClassifier

params = {"criterion": ["gini", "entropy"],             # Criterion to evaluate the purity.
         "max_depth": [7, 9, 11],                           # Maximum depth of the tree
         "min_samples_split": [8, 12, 16]}                   # Stop splitting condition.

grid_search_rf = GridSearchCV(estimator=RandomForestClassifier(n_estimators=10, n_jobs=10), param_grid=params, cv= 5) # Number of trees in the forest is 10

# Run the search on oversampled training data samples.
grid_search_rf.fit(X_train_oversampling, y_train_oversampling)  

GridSearchCV(cv=5, estimator=RandomForestClassifier(n_estimators=10, n_jobs=10),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [7, 9, 11],
                         'min_samples_split': [8, 12, 16]})

In [44]:
# Best set of hyper parameters of the Random Forest
grid_search_rf.best_params_

{'criterion': 'entropy', 'max_depth': 7, 'min_samples_split': 16}

In [45]:
# Build a Random Forest model pipeline from the best set of hyper-parameters found
model_rf = RandomForestClassifier(n_estimators=10, random_state=1, criterion=grid_search_rf.best_params_['criterion'], max_depth=grid_search_rf.best_params_['max_depth'], min_samples_split=grid_search_rf.best_params_['min_samples_split'])     # Initialize the RandomForest

In [46]:
# Train the Random Forest model
model_rf.fit(X_train_oversampling, y_train_oversampling)

RandomForestClassifier(criterion='entropy', max_depth=7, min_samples_split=16,
                       n_estimators=10, random_state=1)

In [89]:
# Calculate some evaluation metrics: precision, recall, f1-score. 'Support' is the number of actual occurrences of the class in the dataset.
from sklearn.metrics import classification_report

# Make prediction on the original test set (after training on the over-sampled training set).
pred_y = model_rf.predict(X_test)

print(classification_report(y_test, pred_y))


              precision    recall  f1-score   support

           0       0.95      0.97      0.96        62
           1       0.96      0.94      0.95        50

    accuracy                           0.96       112
   macro avg       0.96      0.95      0.95       112
weighted avg       0.96      0.96      0.96       112



# COMPARATION

According to the classification report, the models' accuracy on the dataset are: 

1. SVM: 0.94

2. Logistics Regression: 0.96

3. Naive Bayes: 0.94

4. Decision Tree: 0.92

5. Random Forest: 0.96 

That means Random Forest Model can give us the best accuracy of prediction among 5 models